File: netlify/functions/offers.ts
// /api/offers  ·  /api/offers/:id   (admin CRUD + required skills)
import type { Context } from "@netlify/functions";
import { admin } from "./_shared/supabase";
import { requireStaff } from "./_shared/auth";
import { getOrCreateSkill } from "./_shared/db";
import { json, fail, noContent, readBody } from "./_shared/http";

export const config = { path: ["/api/offers", "/api/offers/:id"] };

const SELECT =
  "id, department_id, title, description, field, slots, min_education_level, status, created_at, offer_skills(weight, skill:skills(name))";

function serialize(o: any) {
  return {
    id: o.id,
    department_id: o.department_id,
    title: o.title,
    description: o.description,
    field: o.field,
    slots: o.slots,
    min_education_level: o.min_education_level,
    status: o.status,
    skills: (o.offer_skills ?? [])
      .map((os: any) => ({ name: os.skill?.name, weight: os.weight }))
      .filter((s: any) => s.name),
  };
}

async function applySkills(offerId: number, skills: { name: string; weight?: number }[]) {
  const sb = admin();
  await sb.from("offer_skills").delete().eq("offer_id", offerId);
  for (const ref of skills ?? []) {
    const skillId = await getOrCreateSkill(ref.name);
    await sb
      .from("offer_skills")
      .upsert(
        { offer_id: offerId, skill_id: skillId, weight: ref.weight ?? 1.0 },
        { onConflict: "offer_id,skill_id" },
      );
  }
}

export default async (req: Request, context: Context): Promise<Response> => {
  const user = await requireStaff(req);
  if (user instanceof Response) return user;
  const id = context.params.id ? Number(context.params.id) : null;
  const sb = admin();

  if (req.method === "GET" && !id) {
    const { data, error } = await sb.from("internship_offers").select(SELECT).order("created_at", { ascending: false });
    return error ? fail(error.message, 500) : json((data ?? []).map(serialize));
  }

  if (req.method === "POST") {
    const b = await readBody(req);
    const { data, error } = await sb
      .from("internship_offers")
      .insert({
        department_id: b.department_id,
        title: b.title,
        description: b.description ?? null,
        field: b.field ?? null,
        slots: b.slots ?? 1,
        min_education_level: b.min_education_level ?? null,
        status: b.status ?? "open",
      })
      .select("id")
      .single();
    if (error) return fail(error.message, 500);
    await applySkills(data.id, b.skills ?? []);
    const { data: full } = await sb.from("internship_offers").select(SELECT).eq("id", data.id).single();
    return json(serialize(full), 201);
  }

  if (req.method === "PATCH" && id) {
    const b = await readBody(req);
    const { skills, ...fields } = b;
    if (Object.keys(fields).length) await sb.from("internship_offers").update(fields).eq("id", id);
    if (Array.isArray(skills)) await applySkills(id, skills);
    const { data: full } = await sb.from("internship_offers").select(SELECT).eq("id", id).maybeSingle();
    return full ? json(serialize(full)) : fail("Offre introuvable", 404);
  }

  if (req.method === "DELETE" && id) {
    const { count } = await sb
      .from("applications")
      .select("id", { count: "exact", head: true })
      .eq("offer_id", id);
    if ((count ?? 0) > 0) {
      return fail(`Impossible de supprimer : ${count} candidature(s) liée(s) à cette offre.`, 409);
    }
    const { error } = await sb.from("internship_offers").delete().eq("id", id);
    return error ? fail(error.message, 500) : noContent();
  }

  return fail("Méthode non autorisée", 405);
};

File: backend/app/api/v1/offers.py
"""Internship offer endpoints (CRUD + required skills + embedding)."""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import func, select
from sqlalchemy.orm import Session, selectinload

from app.api.deps import get_current_user, require_staff
from app.core.database import get_db
from app.crud.skill import get_or_create_skill
from app.models.application import Application
from app.models.department import Department
from app.models.offer import InternshipOffer, OfferSkill
from app.models.user import User
from app.schemas.offer import OfferCreate, OfferOut, OfferUpdate
from app.schemas.skill import SkillRef
from app.services.nlp.embeddings import embed_text
from app.services.nlp.pipeline import build_offer_text

router = APIRouter(prefix="/offers", tags=["offers"])


def _serialize(offer: InternshipOffer) -> OfferOut:
    """Convert an ORM offer into its API representation (with skill refs)."""
    data = OfferOut.model_validate(offer)
    data.skills = [SkillRef(name=os.skill.name, weight=os.weight) for os in offer.required_skills]
    return data


def _apply_skills(db: Session, offer: InternshipOffer, skills: list[SkillRef]) -> None:
    """Replace an offer's required skills and refresh its semantic embedding."""
    offer.required_skills.clear()
    db.flush()
    skill_names: list[str] = []
    for ref in skills:
        skill = get_or_create_skill(db, ref.name)
        offer.required_skills.append(OfferSkill(skill_id=skill.id, weight=ref.weight))
        skill_names.append(ref.name)
    # Embed a composite of title/description/field/skills for semantic matching.
    text = build_offer_text(offer.title, offer.description, offer.field, skill_names)
    offer.embedding = embed_text(text)


@router.get("", response_model=list[OfferOut])
def list_offers(
    department_id: int | None = None,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    stmt = (
        select(InternshipOffer)
        .options(selectinload(InternshipOffer.required_skills))
        .order_by(InternshipOffer.created_at.desc())
    )
    if department_id:
        stmt = stmt.where(InternshipOffer.department_id == department_id)
    return [_serialize(o) for o in db.scalars(stmt).all()]


@router.post("", response_model=OfferOut, status_code=201)
def create_offer(
    body: OfferCreate,
    db: Session = Depends(get_db),
    _: User = Depends(require_staff),
):
    if not db.get(Department, body.department_id):
        raise HTTPException(status_code=404, detail="Département introuvable")
    payload = body.model_dump(exclude={"skills"})
    offer = InternshipOffer(**payload)
    db.add(offer)
    db.flush()
    _apply_skills(db, offer, body.skills)
    db.commit()
    db.refresh(offer)
    return _serialize(offer)


@router.get("/{offer_id}", response_model=OfferOut)
def get_offer(offer_id: int, db: Session = Depends(get_db), _: User = Depends(get_current_user)):
    offer = db.get(InternshipOffer, offer_id)
    if not offer:
        raise HTTPException(status_code=404, detail="Offre introuvable")
    return _serialize(offer)


@router.patch("/{offer_id}", response_model=OfferOut)
def update_offer(
    offer_id: int,
    body: OfferUpdate,
    db: Session = Depends(get_db),
    _: User = Depends(require_staff),
):
    offer = db.get(InternshipOffer, offer_id)
    if not offer:
        raise HTTPException(status_code=404, detail="Offre introuvable")
    data = body.model_dump(exclude_unset=True, exclude={"skills"})
    for key, value in data.items():
        setattr(offer, key, value)
    if body.skills is not None:
        _apply_skills(db, offer, body.skills)
    db.commit()
    db.refresh(offer)
    return _serialize(offer)


@router.delete("/{offer_id}", status_code=204)
def delete_offer(offer_id: int, db: Session = Depends(get_db), _: User = Depends(require_staff)):
    offer = db.get(InternshipOffer, offer_id)
    if not offer:
        raise HTTPException(status_code=404, detail="Offre introuvable")
    # Referential integrity: block deletion while applications reference it.
    app_count = db.scalar(
        select(func.count(Application.id)).where(Application.offer_id == offer_id)
    )
    if app_count:
        raise HTTPException(
            status_code=409,
            detail=(
                f"Impossible de supprimer : {app_count} candidature(s) liée(s) " "à cette offre."
            ),
        )
    db.delete(offer)
    db.commit()
